### PR TITLE
[🚧] Completely Refactor IsoAutio

### PR DESCRIPTION
**Pull request by natmfat**
## Summary
Breaking changes
- OOP to easily extend to more models
- Code Splitting
- Better design for future packages (callbacks to hook into features)

This is a massive PR so I wanted to outline what actually changed in detail. I would have included fewer changes but structurally changing the entire library necessitates bigger ones.

1. I organized the 1000+ line file into smaller ones for better maintainability 
2. Changed "isolation_parameters" into a config classes for better typing & refined control (we should try to limit random variables influencing functions as much as possible - I couldn't avoid this in the threshold function but library users won't use it anyway so it should be fine)
3. Created a `BaseAdapter` so all models can just extend from it and generate more easily
4. Updated all docstrings (not true YET)
5. Refactored isolation methods to share internal methods and variables (I need tests cases and expected cases to ensure behavior did not change in this regard)

Here's how I envision people using this aspect of the library. (I will update the tutorial in a future PR)

```python
from PyHa.IsoAutio.adapters import TweetyNetAdapter
from PyHa.IsoAutio import BaseConfig

# we get autocomplete (in development & library usage) because BaseConfig is a class with parameters, not a dictionary
# in this case we leave config as the sensible defaults
tweety_net = TweetyNetAdapter(BaseConfig(audio_dir="TEST")) 
tweety_net.generate(isolate_steinberg_factory(window_size=10)) # still experimenting with this design idea

# the use of callbacks makes it possible to define a custom isolation function
def isolate(result):
    thresh = custom_threshold(result.local_scores)
    return create_entry()
   
tweety_net.generate(isolate)
```

Notice how isolation specific parameters can now be passed *directly* without polluting the configuration. This also means library users can "hook" into more functionality and extend the library by using the Adapters & callbacks. 

# WARNING: THIS IS COMPLETELY UNTESTED AND INCLUDES MANY BREAKING CHANGES.

### CODE REVIEW

**Comment by JacobGlennAyers** [comment]
The entirety of PyHa was purposefully written in a procedural manner so that it fit into the Collaborators' R workflow. I also don't follow the "config class" changes you made since you didn't demo it. Do not push this to main unless the Collaborators agree with it and it has been tested on terabytes of audio data.

**Comment by natmfat** [comment]
> The entirety of PyHa was purposefully written in a procedural manner so that it fit into the Collaborators' R workflow. 

I didn't know that but I see where the function-only approach is coming from. Regardless, I think the project could benefit from major refactoring before new optimization/documentation is added - it just doesn't make sense to expand an already 1000+ line file. Closures could fulfill a similar pattern while still being equally maintainable (or wrapper functions around classes). 

```python
# more functional?
tweety_net = tweety_net_factory(base_config(audio_dir="TEST"))
tweety_net(isolate_steinberg_factory(window_size=10))
```

> I also don't follow the "config class" changes you made since you didn't demo it. 
It is almost exactly the same as the current "isolation parameters" but it was redesigned 
1. to support shared fields (and with correct typing)
2. with the idea that method specific arguments should only be used with those methods

Just as an example:
```python
# this is good
BaseConfig(audio_dir="TEST", manual_id="bird")

# using "window_size" when it only applies to a single method reduces clarity and it infinitely harder to maintain
# basically the equivalent of magic numbers (and we don't really know because dictionaries aren't typed)
{ "window_size": 10, "audio_dir": TEST }

# I think it would be wiser to just pass in window_size when we use it
tweety_net(isolate_steinberg_factory(window_size=10)) # this is an example of that
```
> Do not push this to main unless the Collaborators agree with it and it has been tested on terabytes of audio data.
I definitely agree - this needs extensive review.

**Comment by JacobGlennAyers** [comment]
When we brought this up at last week's collaborator meeting, they also took issue with calling it a "refactor" as they believe a refactor constitutes that the original functionality still works. This in their eyes is more of a rework.

**Comment by natmfat** [comment]
I'm splitting this up into more processable sizes that don't change the underlying infrastructure. 
I intended for this to be a proof of concept (it is and will remain a draft) but I will close for clarity.
